src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+            }
+            catch (Exception e)
+            {
+                act.Context.Logger.LogError(
+                    e,
+                    "Failed completing activity task with completion {Completion}",
+                    completion);
+            }
+            finally
+            {
+                // Remove from running activities
+                runningActivities.TryRemove(tsk.TaskToken, out _);
+            }
+        }
+
+        private async Task<Bridge.Api.ActivityTaskCompletion> ExecuteActivityInternalAsync(
+            RunningActivity act,
+            Bridge.Api.ActivityTask.ActivityTask tsk)
+        {
+            act.Context.Logger.LogDebug(
+                "Running activity {ActivityType} (Token {TaskToken})",
+                act.Context.Info.ActivityType,
+                act.Context.TaskToken);
+            // Completion to be sent back at end of activity
+            var completion = new Bridge.Api.ActivityTaskCompletion()
+            {
+                TaskToken = tsk.TaskToken,
+                Result = new(),
+            };
+            // Set context
+            ActivityContext.AsyncLocalCurrent.Value = act.Context;
+            try
+            {
+                // Find activity or fail
+                if (!activities.TryGetValue(act.Context.Info.ActivityType, out var defn))
+                {
+                    var avail = activities.Keys.ToList();
+                    avail.Sort();
+                    var availStr = string.Join(", ", avail);
+                    throw new ApplicationFailureException(
+                        $"Activity {act.Context.Info.ActivityType} is not registered on this worker," +
+                        $" available activities: {availStr}",
+                        errorType: "NotFoundError");
+                }
+
+                // Deserialize arguments. If the input is less than the required parameter count, we
+                // error.
+                var paramInfos = defn.Delegate.Method.GetParameters();
+                if (tsk.Start.Input.Count < paramInfos.Length &&
+                    !paramInfos[tsk.Start.Input.Count].HasDefaultValue)
+                {
+                    throw new ApplicationFailureException(
+                        $"Activity {act.Context.Info.ActivityType} given {tsk.Start.Input.Count} parameter(s)," +
+                        " but more than that are required by the signature");
+                }
+                // Zip the params and input and then decode each. It is intentional that we discard
+                // extra input arguments that the signature doesn't accept.
+                var paramVals = new List<object?>(tsk.Start.Input.Count);
+                try
+                {
+                    foreach (var (input, paramInfo) in tsk.Start.Input.Zip(paramInfos, (a, b) => (a, b)))
+                    {
+                        var paramVal = await worker.Client.Options.DataConverter.ToValueAsync(
```

Comment:
Prefer to do this concurrently for all params.

Reply:
Fixed